include/zephyr/sys/util.h

```diff
@@ -783,6 +784,32 @@ static inline void mem_xor_128(uint8_t dst[16], const uint8_t src1[16], const ui
 	mem_xor_n(dst, src1, src2, 16);
 }
 
+/**
+ * @brief Compare memory areas and their length
+ *
+ * If the length are 0, return true.
+ *
+ * @param m1 First memory area to compare, cannot be NULL even if length is 0
+ * @param len1 Length of the first memory area to compare
+ * @param m2 Second memory area to compare, cannot be NULL even if length is 0
+ * @param len2 Length of the second memory area to compare
+ *
+ * @returns true if both the length of the memory areas and their content are
+ * equal else false
+ */
+static inline bool memeq(const void *m1, size_t len1, const void *m2, size_t len2)
+{
+	int ret;
+
+	if (len1 != len2) {
+		return false;
+	}
```

Comment:
I don't understand the purpose of the 2nd `len` field. If the caller knows that they 2 inputs are different lengths, why would they even call this function?

My suggestion would be to remove the 2nd `len` argument and just make it 
```
static inline bool memeq(const void *m1, const void *m2, size_t len)
{
    return memcmp(m1, m2, len) == 0;
}

```

I also believe that would make it easier for the compiler to optimize, but without having any data to support that claim.

Reply:
Comparing the length of the input before doing a `memcmp` is a common operation. The goal of that function is about not having to do the comparison yourself as it's common. I don't really see the gain of just doing a `memcmp` that return a boolean to be honnest

Reply:
> I don't really see the gain of just doing a `memcmp` that return a boolean to be honnest

That's also the reason why the function doesn't exist in the standard library :D 

> Comparing the length of the input before doing a memcmp is a common operation. 

Sure, it's common, but not _that_ common. Looking over the uses of `memcmp.* == 0` seems to show that less than 10% of uses does a size comparison first. The most common usecase for this is comparing 2 statically sized values, and adding the len check, especially for an inline function, significantly increases memory use. 

If we move this to be a non-inline function, I'd be more OK with leaving the length check in, but it does still add a check that's mostly not necessary

Reply:
> That's also the reason why the function doesn't exist in the standard library :D

You have a point, maybe I could add both function, would you be fine with that?

> The most common usecase for this is comparing 2 statically sized values, and adding the len check, especially for an inline function, significantly increases memory use.

I understand your concern but I am pretty sure that the compiler will not really inline it (at least from my test with GCC I can tell you that the function is not inlined). Really I put it inline because this is how it's done in this file and without it the compiler complain that it's not used 😅

Reply:
> I understand your concern but I am pretty sure that the compiler will not really inline it (at least from my test with GCC I can tell you that the function is not inlined). Really I put it inline because this is how it's done in this file and without it the compiler complain that it's not used

Might be true :D I've given up trying to follow how compilers do things these days :P 

> You have a point, maybe I could add both function, would you be fine with that?

I guess :) I will defer that to the maintainers of `sys`